src/type-data-frame.c

```diff
@@ -0,0 +1,57 @@
+#include "vctrs.h"
+#include "utils.h"
+
+
+// [[ include("utils.h") ]]
+SEXP new_data_frame(SEXP x, R_len_t n) {
+  x = PROTECT(r_maybe_duplicate(x));
+
+  init_data_frame(x, n);
+
+  if (Rf_length(x) == 0) {
+    Rf_setAttrib(x, R_NamesSymbol, vctrs_shared_empty_chr);
+  }
+
+  UNPROTECT(1);
+  return x;
+}
+
+// [[ include("utils.h") ]]
+bool is_compact_rownames(SEXP x) {
```

Comment:
I feel like these probably should have some `dataframe_` prefix, but it's not that important (and orthogonal to this PR)